### PR TITLE
i#2163: debug register breakpoints

### PR DESCRIPTION
**Pull request by Simorfo**
Adds fields to dcontext structure to follow debug registers.
Their value is set with NtContinue.
When execution gets to an address stored in a debug register,
forges a single step exception.

Fixes i#2163

### CODE REVIEW

**Comment by Simorfo** [comment]
This is going to be a consequent patch.
Before getting the details fine, I would like to know if the general method is alright : 
* store debug register values in context structure (is this the right place ?)
* set them in NtContinue
* check for them in mangle_bb_ilist (right timing ?)
* using dr_insert_clean_call (right function ?) to forge an exception before we execute instruction

**Comment by derekbruening** [comment]
Agreed, if you could elaborate on what is being handled here?  I can imagine a number of problems if an app as part of its normal execution uses debug registers that point at instructions.  If the app sets a debug reg watchpoint on an app PC, and we've already put a block with the corresponding code in our cache, the watchpoint will never trigger.

**Comment by Simorfo** [comment]
>  If the app sets a debug reg watchpoint on an app PC, and we've already put a block with the corresponding code in our cache, the watchpoint will never trigger.

Yes, that is what is happening.
This is an anti-debugging technique.

Here are a few pointers :
section 7 in https://www.symantec.com/connect/articles/windows-anti-debug-reference
section 4.4 in http://piotrbania.com/all/articles/pbania-dbi-unpacking2009.pdf

**Comment by derekbruening** [comment]
You mean, the NtContinue mentioned here is not from a debug register fault that you've assumed has been hit, but from the app using an exception to set the debug registers.  Comments explaining this would help.

If the block is already in the cache, mangle_bb_ilist will not be called again.  The block has to be flushed from the cache.

We'd probably prefer a special exit back to DR to handle it, rather than a clean call which is more of a tool mechanism and complicates flushing.

**Comment by Simorfo** [comment]
> You mean, the NtContinue mentioned here is not from a debug register fault that you've assumed has been hit, but from the app using an exception to set the debug registers.

Yes. But I do not know if there are other ways than using an exception to go to NtContinue.
And SetThreadContext might be used to set the debug registers as well as NtContinue.

So, what should be done ?
How to do a special exit to DR to handle it ?
How to flush if the block is in the cache ?

**Comment by zhaoqin** [comment]
You could terminate a basic block early once see an instruction at breakpoint.
You can reference how DR handles an invalid instruction, i.e., [bb_process_invalid_instr](https://github.com/DynamoRIO/dynamorio/blob/cd4418fbc60c926cca4ed425ffff083c98cef719/core/arch/interp.c#L820), and do something similar.

To flush a block in the code cache, depending on how synchronized flush you want, you can reference dr_delete_fragment, dr_flush_region(), dr_unlink_flush_region(), and dr_delay_flush_region(). They delete code fragment at different time with different restrictions.

**Comment by Simorfo** [comment]
So, I will try to sum it up : 
- store debug register values in context structure -> OK
- set them in NtContinue -> OK but add flushing of fragments (with the debug registers adresses) from the cache 
- check for them in mangle_bb_ilist -> OK
- using dr_insert_clean_call to forge an exception before we execute instruction -> NOT OK, do more something like bb_process_invalid_instr

Is that right ?

**Comment by zhaoqin** [comment]
On Mon, Mar 27, 2017 at 2:53 AM, Simorfo <notifications@github.com> wrote:

> So, I will try to sum it up :
>
>    - store debug register values in context structure -> OK
>
> Is debug register per-thread or per-process?  drcontext is a per-thread
data structure. If the debug register is per-process, it might not be the
best idea to use drcontext.

>
>    - set them in NtContinue -> OK but add flushing of fragments (with the
>    debug registers adresses) from the cache
>
> Is this necessary?
Setting them in NtContinue would affect the execution, e.g., the address
might be touched by a memory read and a real exception would be triggered,
how would we handle that?
Other than the specific case we target here, we need to think of other
common cases, e.g., when the program is under debugging.
If just for this specific case, maybe some global data structure with
proper checking at different places would be better.

**Comment by derekbruening** [comment]
> check for them in mangle_bb_ilist -> OK
> using dr_insert_clean_call to forge an exception before we execute instruction -> NOT OK, do more something like bb_process_invalid_instr

Generally when there are particular instruction addresses that should be handled we look for them during bb decoding, as in check_for_stopping_point(), but in the past they've always been bb start points, so I suppose doing this during mangling makes sense if we don't isolate.  However, mangle_bb_ilist() doesn't seem the best place: better to create a new mangle_* handler invoked by mangle().  But, see Qin's point: if we handle this like probably the closest thing, an invalid instr, which I would also vote for, it would likely not involve mangle at all and instead would isolate into a single-instr bb.

**Comment by Simorfo** [comment]
>  Is debug register per-thread or per-process?  drcontext is a per-thread data structure. If the debug register is per-process, it might not be the best idea to use drcontext.

It is per process, or even per system. Debug register Dr6 has flags to indicate if it should be local to process or system wide.
What structure is better suited then ?

> Setting them in NtContinue would affect the execution

I meant setting the values in the context structure, not the real debug registers value.
I totally agree that the patch should work for other cases
I think that this patch should help, because the example of telock packer is doing just that : debugging itself.

**Comment by zhaoqin** [comment]
On Wed, Mar 29, 2017 at 3:55 AM, Simorfo <notifications@github.com> wrote:

> Is debug register per-thread or per-process? drcontext is a per-thread
> data structure. If the debug register is per-process, it might not be the
> best idea to use drcontext.
>
> It is per process, or even per system. Debug register Dr6 has flags to
> indicate if it should be local to process or system wide.
>
Interesting, shouldn't the OS decide which context switch to swap the debug
register value?

> What structure is better suited then ?
>
Depends on how widely you want to apply it.
If you want to make it cross-platform, somewhere in core/arch/, if x86
only, somewhere core/arch/x86, if Windows only, maybe core/win32.
I cannot think of any existing data structure is good for extending, as I
do not remember we have any data structure to keep track of such process
state, so you may need to create a new one.


> Setting them in NtContinue would affect the execution
>
> I meant setting the values in the context structure, not the real debug
> registers value.
>
Sorry I thought you mean the CONTEXT structure used by NtContinue.

**Comment by Simorfo** [comment]
Ok, I am coming back to this after some more experience with the single step exception issue.

Si, I will try to : 
- store debug register values in a global variable (only for X86)
- set them in NtContinue, and flushing the fragments which are concerned
- check for them in build_bb_ilist, and add a special exit at that point
- forge an exception before when dispatching from the special exit

**Comment by Simorfo** [comment]
By the way, I will reuse code from #2295 where there is already the single step exception generation. 
That is to say : in `build_bb_ilist`, I will only have to set `single_step_addr` and the rest of #2295 should do the magic.

So I think the best plan is to 
- get #2295 merged to master
- rebase this pull request on the new master
- push my changes

**Comment by Simorfo** [comment]
After merge from #2295, here is a new patch.
I am waiting for your review

> store debug register values in a global variable (only for X86)

Is it ok in core/arch/x86/opnd.h as a global variable ?
I did not find a better place

> set them in NtContinue, and flushing the fragments which are concerned

That should be ok. 
There remains to be done Dr6 and Dr7 registers testing : they have flags which tell what to to with Dr1..Dr4.

> check for them in build_bb_ilist, and add a special exit at that point
> forge an exception when dispatching from the special exit

I reused the code from single step exception.

**Comment by Simorfo** [comment]
I had a tests failure because of .data protection
I hope I used correctly the DECLARE_NEVERPROT_VAR

I did not manage to set debug register on 64 bit systems. Exception dispatcher does not call ntcontinue (but rtlrestorecontext)

**Comment by Simorfo** [comment]
I got the test working by setting translation for instruction where a single step is bound to happen.
I hope this is correct

File: suite/tests/security-win32/except-debugreg.c [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+void single_step_addr(void);
+
+
+/* top-level exception handler */
+static LONG
+our_top_handler(struct _EXCEPTION_POINTERS * pExceptionInfo)
+{
+    if (pExceptionInfo->ExceptionRecord->ExceptionCode == EXCEPTION_BREAKPOINT) {
+        //should get here with some of the nops
+        pExceptionInfo->ContextRecord->Dr0 = (unsigned int) single_step_addr;
```
Comment (zhaoqin) [inline suite/tests/security-win32/except-debugreg.c]: more comment about what we are doing here would be more helpful.

File: suite/tests/security-win32/except-debugreg.c [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        pExceptionInfo->ContextRecord->Dr7 = 0x00000101;
+        print("set debug register\n");
+        //goto next instruction to avoid infinite loop
+#ifndef X64
+        pExceptionInfo->ContextRecord->Eip++;
+#else
+        pExceptionInfo->ContextRecord->Rip++;
+#endif
+        return EXCEPTION_CONTINUE_EXECUTION;
+    }
```
Comment (zhaoqin) [inline suite/tests/security-win32/except-debugreg.c]: nit, style guide does not specify this, but I usually prefer
} else if
on the same line unless case like macro breaks it.

File: suite/tests/security-win32/except-debugreg.c [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    else if (pExceptionInfo->ExceptionRecord->ExceptionCode == EXCEPTION_SINGLE_STEP) {
+        count++;
+        print("single step seen\n");
```
Comment (zhaoqin) [inline suite/tests/security-win32/except-debugreg.c]: How do I know this is fired from the code cache at the right location?
Should not you check the Dr0 or pc or something to see it is not other exceptions?

File: suite/tests/CMakeLists.txt [hunk 4]
```diff
@@ -3013,6 +3013,10 @@ else (UNIX)
   tobuild(security-win32.except-guard security-win32/except-guard.c)
   tobuild(security-win32.thread-setcontext security-win32/thread-setcontext.c)
   tobuild(security-win32.singlestep security-win32/singlestep.c)
+  if (NOT X64)
+    # no way to set debug register in 64 bit
+    tobuild(security-win32.except-debugreg security-win32/except-debugreg.c)
+  endif (NOT X64)
```
Comment (zhaoqin) [inline suite/tests/CMakeLists.txt]: so all the code above was for 32 bit only, but most code does not have the conditional macro except the one that checks debug register value.
Why not let that one return false on 64-bit and add comment to explain why we did not do x64. It seems more consistent.

Comment (Simorfo) [inline suite/tests/CMakeLists.txt]: Well, there is indeed a weakness here but I did not know how to fix it :
Most of the code works on 64-bit as well.
The only part that is 32-bit only is the test code setting the debug registers.
I did not find how to get this behavior on 64 bit.

So, what should be done ?

Comment (zhaoqin) [inline suite/tests/CMakeLists.txt]: If it works on x64, could you add comment here about how to (manually) test it on x64 if possible, at least say the 32-bit limitation is test only.
If it does not work on x64, could you return false in debug_register_fire_on_addr for x64, and add comment in that function about x64 not working.

debug_register_fire_on_addr()
{
#ifndef X64
...
#endif
return false;
}

File: core/win32/callback.c [hunk 5]
```diff
@@ -3774,6 +3776,32 @@ intercept_nt_continue(CONTEXT *cxt, int flag)
         LOG(THREAD, LOG_ASYNCH, 3, "target context:\n");
         DOLOG(3, LOG_ASYNCH, { dump_context_info(cxt, THREAD, true); });
 
+        /* Updates debug register values.
+         * FIXME should check dr7 upper bits, and maybe dr6
```
Comment (Simorfo) [inline core/win32/callback.c]: My code just checks the lower bits of Dr7, the upper bits should be checked as well

File: core/arch/interp.c [hunk 6]
```diff
@@ -3494,6 +3510,12 @@ build_bb_ilist(dcontext_t *dcontext, build_bb_t *bb)
             if (!instr_valid(bb->instr))
                 break; /* before eflags analysis! */
 
+#ifdef X86
+            if (my_dcontext != NULL && debug_register_fire_on_addr(bb->cur_pc)) {
```
Comment (Simorfo) [inline core/arch/interp.c]: That was a bug spotted thanks to the test checking the address of the single step exception.
With a debug register, the single step exception happens before execution of the instruction at the address.

Therefore, I check bb->cur_pc to see if next instruction will fire a debug register, in which case a single step exception will be generated right after current instruction.

**Comment by Simorfo** [comment]
I guess I did every suggested change.
What remains to be done for this pull request ?

**Pull request merged**